### ISSUE TITLE
Pods Template ‘Pod Reference’ is Over traversing relationships resulting in ‘stalled’ Template Editor

### ISSUE THREAD

**Issue by PodsBot** [pods-framework/pods#3370]
Pods Template ‘Pod Reference’ is Over traversing relationships resulting in ‘stalled’ Template Editor submitted via Slack by jimtrue

This is one of those 'sometimes' Edge Case, but definitely an issue where the Pods Reference of Fields the Template Editor creates is 'over traversed' to the point of having 4 to 5 or more levels of traversal.

This results in an 'unusable' Pods Reference because no one could possibly want to go that deep into traversal, but also fields like this:

related_tutorials.related_faqs.related_videos.related_faq.related_code.contributors.user_url

In the above Tutorials has a relationship into Faqs and Faqs also has a Relationship into Videos and into Code, but that doesn't means we need to be able to walk across all of those.

**Comment by pglewis** [comment]
Check `pq_loadpod`

**Comment by jimtrue** [comment]
Just create a relationship to a Pod that has a relationship to another Pod that in turn has a relationship to another Pod and you’ll see the above issue

**Comment by pglewis** [comment]
Likely tied-in or the same thing as #3241

**Comment by HughP** [comment]
Funny about the 'edge case' terminology. I ran into the recursive PODs reference issue two days ago with PODs on a site I am building. I didn't think anything of it, rather than it was an un-refined UI issue. -- that is until I added several more PODs and some more relationships...and got a memory error. 
`Fatal error: Allowed memory size of 536870912 bytes exhausted (tried to allocate 72 bytes) in //wordpress/wp-content/plugins/pods/components/Templates/includes/functions-pod_reference.php on line 89`

I can still use the site, just not the templates admin UI. (which makes me think that the PODs reference recursion is causing the memory error) I am only running 12 PODs (some CPTs and some Custom Taxonomies, and one each are extending Users and Media).

**Comment by quasel** [comment]
Can you share your POD's configuration? -> POD's admin -> Migrate Packages (maby you need to enable it under Components first!

**Comment by jimtrue** [comment]
#3241 includes a fix for this. This is a known issue when you have more than 3 related bidirectional fields connecting to either itself (self relating) or multiple pods relating to the same Pods. It just creates an over referenced traversal field list.

**Comment by HughP** [comment]
@jimtrue I definitely have more than three related fields. Assuming that #3241 does fix the issue, is there a way to keep this code intact or will I need to readdress this issue with every pods update? The update from the 15th did not address this as I understand it.

@quasel Do you still want the pods configuration? I am willing to find and replace as #3241 suggests.

**Comment by quasel** [comment]
@HughP well if it fixes it for you no need for further investigation ^^ just keep an eye on it until it's shipped with a release ;)

**Comment by pglewis** [comment]
#3241 suggests a fix, but it isn't a PR or necessarily the best fix IMO.  I think it's possible to solve this if we can detect a circular reference and only traverse those once.  This seems a more "proper" solution to me than an artificial depth limit, which is just treating the symptom.

**Comment by jimtrue** [comment]
Yep, that was my point #3241 suggested a workaround, but definitely not a fix.

**Comment by IndieRafael** [comment]
jimtrue asked me to post here some of the chat on our Slack chat about this:

Problem description:
1) Earlier we chatted about my problem with pods reference recursion. As you suggested, I turned off pods reference. That helped a little, but I still find pods template unusable when editing magic tags on my Linux box, and completely unusable on my Wind7 box. I've tested on Linux Mint and Windows 7 operating systems, with Firefox and and MSIE.  Best result is on Linux box, where speed is normal editing templates UNLESS I start editing a magic tag. Within about 2 seconds, the browser freezes for 10 secons, then repeat the process, so I cannot edit templatges. On Windows box, I cannot edit templates because the editor window doesn't even show up, only the side dashboard and top right status block. (But editor window comes up on regular WP pages.) On MSIE, it freezes the page. I have disabled all but a few essential plugins, cleared browser cache, etc. My customer-friendly commercial host has no ongoing problem. Problem is constant, not intermittent. 

2) To be clear: I'm still finalizing my pods and fields and taxonomies. I really don't want to reduce the number bidirectional fields in my pods. I currently have 6 custom content types with the following respective numbers of bidirectional fields each:  0, 2, 2, 3, 3, 4.  So far, there is little data or linkages actually in the database, just the structure and some test data. I could cut the number of bidirectioal fields if that were the only way to make it work. I'm committed to Pods for at least phase one of this project and don't want to fight the tool. But if this is too much for Pods, I need to start considering alternatives for the next phase. If there are configuration tweaks appropriate for this kind of setup, I'm open to that.  Thanks again for your help earlier.

Diagnosing:
Later, I used Firefox Browser Console, logging Javascript. I see a lot like this:
POST   http://mysiteexample.com/wp-admin/admin-ajax.php      some take almost a full second
POST   http://mysiteexample.com/wp-admin/admin-ajax.php      some take almost a full second
It does this a couple of times per minute, perhaps. There are a few other JS items that repeat, but none that seem related to the Pods template problem.

Workaround:
So set it to "Select Pod to use as reference."  So no pod is selected. 

When I can get that to stick, it seems better. But I have to start over each time I open a template, and even setting it to "Select Pod" doesn't always seem to work from my limited testing.

Finally, Jim True advised I should copy all of my existing templates to new templates, and make sure none of the new templates select a pod type in pod reference.
